packages/server/utils/isRequestToJoinDomainAllowed.ts

```diff
@@ -1,43 +1,73 @@
 import isCompanyDomain from './isCompanyDomain'
 import getRethink from '../database/rethinkDriver'
 import {RDatum} from '../database/stricterR'
+import isUserVerified from './isUserVerified'
+import User from '../database/types/User'
+import {DataLoaderWorker} from '../graphql/graphql'
+import isValid from '../graphql/isValid'
+import TeamMember from '../database/types/TeamMember'
 
 export const getEligibleOrgIdsByDomain = async (
   activeDomain: string,
-  viewerId: string,
-  limit?: number
+  userId: string,
+  dataLoader: DataLoaderWorker
 ) => {
-  const BIG_ENOUGH_LIMIT = 9999
-
   if (!isCompanyDomain(activeDomain)) {
     return []
   }
 
   const r = await getRethink()
 
-  return r
+  const orgs = await r
     .table('Organization')
     .getAll(activeDomain, {index: 'activeDomain'})
-    .filter((org: RDatum) => org('featureFlags').contains('promptToJoinOrg'))
-    .filter((org: RDatum) =>
-      r
+    .merge((org: RDatum) => ({
+      members: r
         .table('OrganizationUser')
         .getAll(org('id'), {index: 'orgId'})
-        .filter({inactive: false, removedAt: null})
+        .orderBy('joinedAt')
         .coerceTo('array')
-        .do((orgUsers: RDatum) =>
-          orgUsers
-            .count()
-            .gt(1)
-            .and(orgUsers.filter((ou: RDatum) => ou('userId').eq(viewerId)).isEmpty())
-        )
+    }))
+    .merge((org: RDatum) => ({
+      founder: org('members').nth(0).default(null),
+      billingLeads: org('members').filter({role: 'BILLING_LEADER', inactive: false}),
+      activeMembers: org('members').filter({inactive: false, removedAt: null}).count()
+    }))
+    .filter((org: RDatum) =>
+      org('activeMembers').gt(0).and(org('members').filter({userId}).isEmpty())
     )
-    .limit(limit ?? BIG_ENOUGH_LIMIT)('id')
     .run()
+
+  const eligibleOrgs = await Promise.all(
+    orgs.map(async (org) => {
+      const {founder} = org
+      const importentMembers = org.billingLeads.slice() as TeamMember[]
+      if (!founder.inactive && !founder.removedAt && founder.role !== 'BILLING_LEADER') {
+        importentMembers.push(founder)
+      }
+
+      const users = (
+        await dataLoader.get('users').loadMany(importentMembers.map(({userId}) => userId))
```

Comment:
feels like we can first iterate through orgs to get all user ids, and then make a single dataloader query

Reply:
Yes, but this would make the code less readable and it wouldn't improve anything. This is the unique benefit of a dataloader, although a bit tricky to wrap your head around. JavaScript is single threaded, so the code gets executed until it hits an unresolved promise/await call, then it's added to a queue. The dataloader [ensures](https://github.com/graphql/dataloader/blob/d336bd15282664e0be4b4a657cb796f09bafbc6b/src/index.js#L213-L238) that the result of `loadMany` will be unresolved and it will be added to the end of the queue. Once it's the dataloaders turn, it will execute all load calls at once and resolve all promises at once.
I think it's easiest to understand with the following small example:
```ts
import DataLoader from 'dataloader'

const testLoader = new DataLoader(async (keys: string[]) => {
  console.log('loading ', keys)
  return keys
})

test('Testing loading queue', async () => {
  await Promise.all([1,2,3].map(async (i) => {
    console.log('mapping ', i)
    await testLoader.load(i)
  })
})
```
will print
```
    mapping  1
    mapping  2
    mapping  3
    loading  [ 1, 2, 3 ]
```